docs/standard/microservices-architecture/multi-container-microservice-net-applications/implementing-api-gateways-with-ocelot.md

```diff
@@ -0,0 +1,566 @@
+
+In any case, the authentication module of Ocelot API Gateway will be visited at first when trying to use any secured microservice (if secured at the API Gateway level). That will re-direct to visit the Identity or auth microservice to get the access token so so you can visit the protected services with the access_token.
+
+The way you secure with authentication any service at the API Gateway level is by setting the AuthenticationProviderKey in its related settings at the configuration.json.
+
+```csharp
+    {
+      "DownstreamPathTemplate": "/api/{version}/{everything}",
+      "DownstreamScheme": "http",
+      "DownstreamHostAndPorts": [
+        {
+          "Host": "basket.api",
+          "Port": 80
+        }
+      ],
+      "UpstreamPathTemplate": "/api/{version}/b/{everything}",
+      "UpstreamHttpMethod": [],
+      "AuthenticationOptions": {
+        "AuthenticationProviderKey": "IdentityApiKey",
+        "AllowedScopes": []
+      }
+    }
+```
+
+When Ocelot runs, it will look at the ReRoutes AuthenticationOptions.AuthenticationProviderKey and check that there is an Authentication Provider registered with the given key. If there isn't, then Ocelot will not start up. If there is, then the ReRoute will use that provider when it executes.
+
+Because the Ocelot WebHost is configured with the `authenticationProviderKey = "IdentityApiKey"`, that will require authentication whenever that service has any requests without any auth token. 
+
+```csharp
+namespace OcelotApiGw
+{
+    public class Startup
+    {
+        private readonly IConfiguration _cfg;
+
+        public Startup(IConfiguration configuration)
+        {
+            _cfg = configuration;
+        }
+        public void ConfigureServices(IServiceCollection services)
+        {
+            var identityUrl = _cfg.GetValue<string>("IdentityUrl");
+            var authenticationProviderKey = "IdentityApiKey";
+                         //…
+            services.AddAuthentication()
+                .AddJwtBearer(authenticationProviderKey, x =>
+                {
+                    x.Authority = identityUrl;
+                    x.RequireHttpsMetadata = false;
+                    x.TokenValidationParameters = new Microsoft.IdentityModel.Tokens.TokenValidationParameters()
+                    {
+                        ValidAudiences = new[] { "orders", "basket", "locations", "marketing", "mobileshoppingagg", "webshoppingagg" }
+                    };                   
+                });
+            //...
+```
+
+Then, you also need to set authorization with the [Authorize] attribute on any resource to be accessed like the microservices, such as in the following Basket microservice controller.
+
+```csharp
+namespace Microsoft.eShopOnContainers.Services.Basket.API.Controllers
+{
+    [Route("api/v1/[controller]")]
+    [Authorize]
+    public class BasketController : Controller
+    {   //...
```

Comment:
github doesn't like when the fences are not closed for code blocks. it breaks formatting too

Reply:
👍